service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceResponseDecoder.java

```diff
@@ -92,6 +92,30 @@ public ServiceResponseDecoder(ObjectMapper objectMapper, JavaType javaType, Java
                     throw new IOException("Response object de-serialization error", e);
                 }
             }
+        } else if (statusCode >= 500 && statusCode <=599) {
+            try {
+                String serviceResponse = StringUtils.convertStreamToString(httpResponse.getEntity().getContent());
+                Object errorResponse = null;
+                try {
+                    if (errorType != null) {
+                        errorResponse = objectMapper.readValue(serviceResponse, errorType);
+                    }
+                } catch (Exception e) {
+                    logger.warn("Error while de-serializing 5XX response to given errorType statusCode:{} exception: {}", statusCodeString, e.getMessage());
+                }
+                logger.warn("5XX response statusCode:{} response: {}", statusCodeString, serviceResponse);
+                Class<? extends ServiceClientException> exceptionClass;
+                if (exceptions.containsKey(statusCodeString))
+                    exceptionClass = exceptions.get(statusCodeString);
+                else
+                    exceptionClass = exceptions.get("default");
+
+                return new ServiceResponse<T>(exceptionClass.getConstructor(String.class, Object.class).newInstance(serviceResponse, errorResponse), headers);
+
+            } catch (Exception e) {
+                logger.error("Error de-serializing 5XX response statusCode:{} exception: {} ", statusCodeString, e.getMessage());
+                throw new IOException("5XX response de-serialization error", e);
```

Comment:
Just logs differ, do we need a separate if block copying the whole code? You can take a look at how this is done in poseidon 5.x (where in addition to differential logging treatment for 5xx and non 5xx, we also treat 5xx as hystrix errors).